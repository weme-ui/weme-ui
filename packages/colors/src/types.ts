export type ColorAppearance = 'light' | 'dark'

export type ColorType = 'accent' | 'neutral'

export type ColorScalesArray<T = string> = [T, T, T, T, T, T, T, T, T, T, T, T]

export type Colors<T = string> = Record<string, ColorScalesArray<T>>
