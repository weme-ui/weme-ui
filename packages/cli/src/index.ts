#!/usr/bin/env node
import { runMain } from 'citty'
import { cli } from './cli'

runMain(cli)
