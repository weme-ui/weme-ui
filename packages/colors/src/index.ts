export * from './colors'
export * from './transform'
export * from './types'

export { getColorNames, toColorString } from './utils'
