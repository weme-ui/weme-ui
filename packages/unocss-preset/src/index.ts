import type { UserPresetOptions } from './types'
import { definePreset, mergeDeep } from 'unocss'
import { resolveOptions } from './options'
import { preflights } from './preflights'
import { shortcuts } from './shortcuts'
import { theme } from './theme'
import { variants } from './variants'

export * from './types/themes'

export const presetHalo = definePreset((userOptions: UserPresetOptions = {}) => {
  const options = resolveOptions(userOptions)

  return {
    name: '@weme-ui/unocss-preset-halo',
    extendTheme: (t) => {
      const {
        colors,
        font,
        text,
        radius,
        breakpoint,
        verticalBreakpoint,
        ...others
      } = theme(options)

      return {
        ...mergeDeep(t, others),
        colors,
        font,
        text,
        radius,
        breakpoint,
        verticalBreakpoint,
      }
    },
    preflights: preflights(options),
    shortcuts,
    variants,
    shorthands: {
      'theme-colors': '(accent|neutral|info|success|warning|error)',
      'scales': '(1|2|3|4|5|6|7|8|9|10|11|12)',
    },
    enforce: 'post',
  }
})
