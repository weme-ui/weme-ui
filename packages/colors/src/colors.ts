import type Color from 'colorjs.io'
import type { ColorAppearance, Colors } from './types'
import { getColorNames, getRadixColor } from './utils'

const lightColors = Object.fromEntries(
  getColorNames().map((name: string) => [
    name,
    getRadixColor('light', name),
  ]),
) as Colors<Color>

const darkColors = Object.fromEntries(
  getColorNames().map((name: string) => [
    name,
    getRadixColor('dark', name),
  ]),
) as Colors<Color>

const lightNeutralColors = Object.fromEntries(
  getColorNames('neutral').map((name: string) => [
    name,
    getRadixColor('light', name),
  ]),
) as Colors<Color>

const darkNeutralColors = Object.fromEntries(
  getColorNames('neutral').map((name: string) => [
    name,
    getRadixColor('dark', name),
  ]),
) as Colors<Color>

export function getColors(appearance: ColorAppearance, isNeutral?: boolean) {
  return isNeutral
    ? appearance === 'light' ? lightNeutralColors : darkNeutralColors
    : appearance === 'light' ? lightColors : darkColors
}
