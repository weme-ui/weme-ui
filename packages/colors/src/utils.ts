import type { ColorAppearance, ColorScalesArray, ColorType } from './types'
import * as RadixColors from '@radix-ui/colors'
import Color from 'colorjs.io'

const SPECIAL_COLOR_NAMES = ['black', 'white'] as const
const NEUTRAL_COLOR_NAMES = ['gray', 'mauve', 'slate', 'sage', 'olive', 'sand'] as const
const ACCENT_COLOR_NAMES = ['tomato', 'red', 'ruby', 'crimson', 'pink', 'plum', 'purple', 'violet', 'iris', 'indigo', 'blue', 'cyan', 'teal', 'jade', 'green', 'grass', 'bronze', 'gold', 'brown', 'orange', 'amber', 'yellow', 'lime', 'mint', 'sky'] as const
const COLOR_NAMES = [...SPECIAL_COLOR_NAMES, ...NEUTRAL_COLOR_NAMES, ...ACCENT_COLOR_NAMES] as const

export type RadixColorNames = keyof typeof RadixColors

export type AllColorNames = typeof NEUTRAL_COLOR_NAMES[number] | typeof ACCENT_COLOR_NAMES[number]
export type AccentColorNames = typeof ACCENT_COLOR_NAMES[number]
export type NeutralColorNames = typeof NEUTRAL_COLOR_NAMES[number]

export function isRadixColorName(color: string) {
  return COLOR_NAMES.includes(color as any)
}

export function getColorNames(type?: ColorType) {
  return !type
    ? COLOR_NAMES
    : type === 'accent'
      ? ACCENT_COLOR_NAMES
      : NEUTRAL_COLOR_NAMES
}

export function getRadixColorName(name: string, appearance: ColorAppearance) {
  return name === 'black'
    ? 'blackP3A'
    : name === 'white'
      ? 'whiteP3A'
      : `${name}${appearance === 'light' ? 'P3' : 'DarkP3'}` as RadixColorNames
}

export function getRadixColor<
  S extends boolean = false,
  T extends S extends true ? string : Color = S extends true ? string : Color,
  RT = ColorScalesArray<T>,
>(
  appearance: ColorAppearance,
  name: string,
  stringify?: boolean,
): RT {
  const colorName = getRadixColorName(name, appearance)

  return Object.values(RadixColors[colorName]).map(
    color => stringify
      ? toColorString(color)
      : new Color(color).to('oklch'),
  ) as RT
}

export function toColorString(colorOrStr: Color | string): string {
  const color = typeof colorOrStr === 'string'
    ? new Color(colorOrStr)
    : colorOrStr

  return color
    .to('oklch')
    .toString({ precision: 4 })
}
