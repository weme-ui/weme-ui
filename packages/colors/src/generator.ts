import type { Colors, ColorScalesArray } from './types'
import BezierEasing from 'bezier-easing'
import Color from 'colorjs.io'
import { getColorNames, toColorString } from './utils'

export function generateColor(
  color: string,
  colors: Colors<Color>,
  background: string,
): ColorScalesArray {
  const sourceColor = new Color(color).to('oklch')
  const backgroundColor = new Color(background).to('oklch')

  const scales = getScaleColors(sourceColor, colors, backgroundColor)

  scales[8] = sourceColor
  scales[9] = getHoverColor(sourceColor, [scales])

  scales[10].coords[1] = Math.min(
    Math.max(scales[8].coords[1], scales[7].coords[1]),
    scales[10].coords[1],
  )

  scales[11].coords[1] = Math.min(
    Math.max(scales[8].coords[1], scales[7].coords[1]),
    scales[11].coords[1],
  )

  return scales.map(toColorString) as ColorScalesArray
}

function getScaleColors(
  source: Color,
  colors: Record<string, ColorScalesArray<Color>>,
  backgroundColor: Color,
): ColorScalesArray<Color> {
  const allColors: { name: string, color: Color, distance: number }[] = []

  Object.entries(colors).forEach(([name, scale]) => {
    for (const color of scale) {
      const distance = source.deltaEOK(color)
      allColors.push({ name, distance, color })
    }
  })

  allColors.sort((a, b) => a.distance - b.distance)

  const closestColors = allColors.filter(
    (color, i, arr) => i === arr.findIndex(value => value.name === color.name),
  )

  const grayColorNames = getColorNames('neutral') as readonly string[]
  const allAreGrays = closestColors.every(color =>
    grayColorNames.includes(color.name),
  )
  if (!allAreGrays && grayColorNames.includes(closestColors[0].name)) {
    while (grayColorNames.includes(closestColors[1].name)) {
      closestColors.splice(1, 1)
    }
  }

  const closestA = closestColors[0]
  const closestB = closestColors[1]

  const a = closestB.distance
  const b = closestA.distance
  const c = closestA.color.deltaEOK(closestB.color)

  const cosA = (b ** 2 + c ** 2 - a ** 2) / (2 * b * c)
  const radA = Math.acos(cosA)
  const sinA = Math.sin(radA)

  const cosB = (a ** 2 + c ** 2 - b ** 2) / (2 * a * c)
  const radB = Math.acos(cosB)
  const sinB = Math.sin(radB)

  const tanC1 = cosA / sinA
  const tanC2 = cosB / sinB
  const ratio = Math.max(0, tanC1 / tanC2) * 0.5

  const colorA = colors[closestA.name]
  const colorB = colors[closestB.name]

  const scale = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11].map(i =>
    new Color(Color.mix(colorA[i], colorB[i], ratio)).to('oklch'),
  ) as ColorScalesArray<Color>

  const baseColor = scale.slice()
    .sort((a, b) => source.deltaEOK(a) - source.deltaEOK(b))[0]

  const ratioC = source.coords[1] / baseColor.coords[1]

  scale.forEach((color) => {
    color.coords[1] = Math.min(
      source.coords[1] * 1.5,
      color.coords[1] * ratioC,
    )
    color.coords[2] = source.coords[2]
  })

  // Light
  if (scale[0].coords[0] > 0.5) {
    const lightnessScale = scale.map(({ coords }) => coords[0])
    const backgroundL = Math.max(0, Math.min(1, backgroundColor.coords[0]))
    const newLightnessScale = transposeProgressionStart(
      backgroundL,
      [1, ...lightnessScale],
      [0, 2, 0, 2] as [number, number, number, number],
    )

    newLightnessScale.shift()

    newLightnessScale.forEach((lightness, i) => {
      scale[i].coords[0] = lightness
    })

    return scale
  }

  // Dark
  const ease: [number, number, number, number] = [1, 0, 1, 0]
  const referenceBackgroundColorL = scale[0].coords[0]
  const backgroundColorL = Math.max(0, Math.min(1, backgroundColor.coords[0]))

  const ratioL = backgroundColorL / referenceBackgroundColorL

  if (ratioL > 1) {
    const maxRatio = 1.5

    for (let i = 0; i < ease.length; i++) {
      const metaRatio = (ratioL - 1) * (maxRatio / (maxRatio - 1))
      ease[i] = ratioL > maxRatio ? 0 : Math.max(0, ease[i] * (1 - metaRatio))
    }
  }

  const lightnessScale = scale.map(({ coords }) => coords[0])
  const newLightnessScale = transposeProgressionStart(
    backgroundColorL,
    lightnessScale,
    ease,
  )

  newLightnessScale.forEach((lightness, i) => {
    scale[i].coords[0] = lightness
  })

  return scale
}

function getHoverColor(source: Color, scales: ColorScalesArray<Color>[]) {
  const [L, C, H] = source.coords
  const newL = L > 0.4 ? L - 0.03 / (L + 0.1) : L + 0.03 / (L + 0.1)
  const newC = L > 0.4 && !Number.isNaN(H) ? C * 0.93 + 0 : C
  const hoverColor = new Color('oklch', [newL, newC, H])

  let closestColor = hoverColor
  let minDistance = Infinity

  scales.forEach((scale) => {
    for (const color of scale) {
      const distance = hoverColor.deltaEOK(color)
      if (distance < minDistance) {
        minDistance = distance
        closestColor = color
      }
    }
  })

  hoverColor.coords[1] = closestColor.coords[1]
  hoverColor.coords[2] = closestColor.coords[2]

  return hoverColor
}

function transposeProgressionStart(
  to: number,
  arr: number[],
  curve: [number, number, number, number],
) {
  return arr.map((n, i, arr) => {
    const lastIndex = arr.length - 1
    const diff = arr[0] - to
    const fn = BezierEasing(...curve)
    return n - diff * fn(1 - i / lastIndex)
  })
}
