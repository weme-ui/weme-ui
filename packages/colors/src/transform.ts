import type { ColorAppearance, ColorScalesArray } from './types'
import { getColors } from './colors'
import { generateColor } from './generator'
import { getRadixColor, isRadixColorName } from './utils'

interface TransformColorOptions {
  color: string
  appearance?: ColorAppearance
  isNeutral?: boolean
  background?: string
}

export function transformColor({
  color,
  appearance = 'light',
  isNeutral = false,
  background = '#fff',
}: TransformColorOptions): ColorScalesArray {
  if (isRadixColorName(color))
    return getRadixColor(appearance, color, true)

  const colors = getColors(appearance, isNeutral)

  return generateColor(color, colors, background)
}
