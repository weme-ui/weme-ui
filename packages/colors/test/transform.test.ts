import { describe, expect, it } from 'vitest'
import { transformColor } from '../src/transform'

describe('transform', () => {
  it('should transform color to scales', () => {
    const colors = [
      transformColor({ appearance: 'light', color: '#05f' }),
      transformColor({ appearance: 'dark', color: 'color(display-p3 0.733 0.864 0.724)' }),
      transformColor({ appearance: 'light', color: 'tomato' }),
      transformColor({ appearance: 'dark', color: 'ruby' }),
    ]

    expect(colors).toMatchSnapshot()
  })
})
