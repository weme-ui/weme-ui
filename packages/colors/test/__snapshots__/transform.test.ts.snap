// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform > should transform color to scales 1`] = `
[
  [
    "oklch(99.41% 0.0023 262.6)",
    "oklch(98.27% 0.011 262.6)",
    "oklch(96.06% 0.0233 262.6)",
    "oklch(93.56% 0.0456 262.6)",
    "oklch(90.37% 0.0686 262.6)",
    "oklch(86.31% 0.0944 262.6)",
    "oklch(80.72% 0.1159 262.6)",
    "oklch(73.09% 0.1525 262.6)",
    "oklch(53.32% 0.2596 262.6)",
    "oklch(48.58% 0.2366 262.6)",
    "oklch(51.8% 0.2366 262.6)",
    "oklch(31.46% 0.1181 262.6)",
  ],
  [
    "oklch(100% 0.011 146.8)",
    "oklch(95.03% 0.0139 146.8)",
    "oklch(93.52% 0.0337 146.8)",
    "oklch(90.82% 0.0495 146.8)",
    "oklch(88.1% 0.057 146.8)",
    "oklch(85.58% 0.0631 146.8)",
    "oklch(83.62% 0.0713 146.8)",
    "oklch(82.28% 0.0828 146.8)",
    "oklch(86.18% 0.1071 146.8)",
    "oklch(83.06% 0.1071 146.8)",
    "oklch(85.36% 0.1071 146.8)",
    "oklch(90.47% 0.0714 146.8)",
  ],
  [
    "oklch(99.35% 0.0031 22.75)",
    "oklch(98.42% 0.0071 31.02)",
    "oklch(95.44% 0.022 30.88)",
    "oklch(92.6% 0.0471 31.76)",
    "oklch(89.21% 0.0626 31.49)",
    "oklch(85.24% 0.077 32.01)",
    "oklch(80.28% 0.0943 32.17)",
    "oklch(74.15% 0.1182 32.07)",
    "oklch(62.7% 0.1937 33.31)",
    "oklch(60.4% 0.1953 33.26)",
    "oklch(56.65% 0.1977 32.81)",
    "oklch(34.61% 0.0798 30.53)",
  ],
  [
    "oklch(18.88% 0.0137 2.904)",
    "oklch(20.77% 0.0159 5.29)",
    "oklch(25.38% 0.0602 6.544)",
    "oklch(29.21% 0.0888 5.505)",
    "oklch(33.27% 0.0997 6.398)",
    "oklch(38.2% 0.105 7.589)",
    "oklch(44.83% 0.1165 9.017)",
    "oklch(54.38% 0.1454 11.06)",
    "oklch(62.84% 0.195 13.19)",
    "oklch(66.35% 0.1794 13.69)",
    "oklch(78.61% 0.1611 17.01)",
    "oklch(90.54% 0.0527 355.8)",
  ],
]
`;
